File: questionThree.c
/*
*Author: Matthew Zheng
*Purpose: Solution to Project Euler Q3: Largest Prime Factor
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define TRUE 1
#define FALSE 0

unsigned long long int ceilingUp(float);
unsigned long long int prime(unsigned long long int);

int main(void){
	unsigned long long int numTest = 600851475143;
	unsigned long long int biggest;
	for (int i = 1; i <= numTest; i++){
		/*Look for the largest multiple*/
		if (numTest % i == 0){
			//printf("%llu ", i);
			/*Check if prime by comparing it to the ceiling value of root*/
			if (prime(i)){
				biggest = i;
			}
			else{
				//puts("There is no prime.");
			}
		}
	}

	printf("The largest prime is: %llu\n", biggest);

	system("PAUSE");
	return 0;
}

unsigned long long int ceilingUp(float num){
	float difference;
	/*Get absolute value difference*/
	if ((((int)num) - num) < 0){
		difference = -(((int)num) - num);
	}
	else{
		difference = ((int)num) - num;
	}
	/*Comparison test to see if round up or down*/
	if (difference < 0.5){
		return(((int)num));
	}
	else if (difference >= 0.5){
		return(((int)num) + 1);
	}
}

unsigned long long int prime(unsigned long long int numCheck){
	/*Declare new variable to act as a switch*/
	int primeNum = FALSE;

	if (numCheck <= 2){
		/*Numbers 2 or under are prime*/
		primeNum = TRUE;
	}
	else{
		for (int i = 2; i < ceilingUp(sqrt(numCheck)); i++){
			if (numCheck % i == 0){
				primeNum = FALSE;
				break;
			}
			else {
				primeNum = TRUE;
			}
		}
	}
	return(primeNum);
}

File: questionThree.py
#! usr/bin/env python
_author_ = "Matthew Zheng"
_purpose_ = "Solution to Project Euler Q3: Largest Prime Factor"

import sys
import math

def main():
    numTest = 600851475143
    #look for the largest multiple
    for i in range (1, numTest):
        if(numTest % i == 0):
            #check if number is prime
            if (prime(i)):
                biggestNum = i

    print("The largest prime is", biggestNum)

def ceilingUp(num):
    #We are not going to use the built in ceiling function on line 24
    #return(math.ceil(num))
    if num-int(num) < 0:
        difference = -(num-int(num))
    else:
        difference = num-int(num)
    if difference < 0.5:
        return(int(num))
    else:
        return(int(num)+1)

def prime(numCheck):
    primeNum = False
    if numCheck <= 2:
        #all number up to and including 2 are prime
        primeNum = True
    else:
        for i in range(2, ceilingUp(math.sqrt(numCheck))):
            if numCheck % i == 0:
                primeNum = False
                break
            else:
                primeNum = True
    return(primeNum)


if __name__ == "__main__":
    main()
